packages/messaging/src/controllers/controller-interface.ts

```diff
@@ -100,26 +108,31 @@ export default class ControllerInterface {
    * 3) If the database cache is invalidated: Send a request to FCM to update
    * the token, and to check if the token is still valid on FCM-side.
    */
-  private manageExistingToken(
-    tokenDetails: Object,
-    swReg: ServiceWorkerRegistration
+  private async manageExistingToken(
+    swReg: ServiceWorkerRegistration,
+    pushSubscription: PushSubscription,
+    publicVapidKey: Uint8Array,
+    tokenDetails: Object
   ): Promise<string> {
-    return this.isTokenStillValid(tokenDetails).then(isValid => {
-      if (isValid) {
-        const now = Date.now();
-        if (now < tokenDetails['createTime'] + TOKEN_EXPIRATION_MILLIS) {
-          return tokenDetails['fcmToken'];
-        } else {
-          return this.updateToken(tokenDetails, swReg);
-        }
+    const isTokenValid = await this.isTokenStillValid(tokenDetails);
+    if (isTokenValid) {
+      const now = Date.now();
+      if (now < tokenDetails['createTime'] + TOKEN_EXPIRATION_MILLIS) {
+        return tokenDetails['fcmToken'];
       } else {
-        // If the VAPID details are updated, delete the existing token,
-        // and create a new one.
-        return this.deleteToken(tokenDetails['fcmToken']).then(() => {
-          return this.getNewToken(swReg);
-        });
+        return this.updateToken(
+          swReg,
+          pushSubscription,
+          publicVapidKey,
+          tokenDetails
+        );
       }
-    });
+    }
+
+    // If the VAPID details are updated, delete the existing token,
```

Comment:
May reword this to as the logic suggests the delete and create steps are undertaken if the token is not longer, like when VAPID Details are changed.

Reply:
done